Userland/Libraries/LibGfx/TGALoader.cpp

```diff
@@ -0,0 +1,248 @@
+    }
+
+    ALWAYS_INLINE i16 read_i16()
+    {
+        return read_i8() | read_i8() << 8;
+    }
+
+    ALWAYS_INLINE u32 read_u32()
+    {
+        return read_u16() | read_u16() << 16;
+    }
+
+    ALWAYS_INLINE i32 read_i32()
+    {
+        return read_i16() | read_i16() << 16;
+    }
+
+    ALWAYS_INLINE TGAPixel read_pixel(u8 bits_per_pixel)
+    {
+        auto pixel = TGAPixel();
+
+        switch (bits_per_pixel) {
+        case 24:
+            pixel.components.blue = read_u8();
+            pixel.components.green = read_u8();
+            pixel.components.red = read_u8();
+            pixel.components.alpha = 0xFF;
+            return pixel;
+
+        case 32:
+            pixel.components.blue = read_u8();
+            pixel.components.green = read_u8();
+            pixel.components.red = read_u8();
+            pixel.components.alpha = read_u8();
+            return pixel;
+
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    size_t index() const
+    {
+        return m_index;
+    }
+
+private:
+    Span<u8 const> m_data;
+    size_t m_index { 0 };
+};
+
+struct TGALoadingContext {
+    TGAHeader header;
+    Span<u8 const> bytes;
+    size_t file_size;
+    OwnPtr<TGAReader> reader = { nullptr };
+    RefPtr<Gfx::Bitmap> bitmap;
+};
+
+TGAImageDecoderPlugin::TGAImageDecoderPlugin(u8 const* file_data, size_t file_size)
+{
+    m_context = make<TGALoadingContext>();
+    m_context->bytes = Span<u8 const>(file_data, file_size);
+    m_context->file_size = move(file_size);
+    m_context->reader = make<TGAReader>(m_context->bytes);
+}
+
+TGAImageDecoderPlugin::~TGAImageDecoderPlugin() = default;
+
+IntSize TGAImageDecoderPlugin::size()
+{
+    return IntSize { m_context->header.width, m_context->header.height };
+}
+
+void TGAImageDecoderPlugin::set_volatile()
+{
+    if (m_context->bitmap)
+        m_context->bitmap->set_volatile();
+}
+
+bool TGAImageDecoderPlugin::set_nonvolatile(bool& was_purged)
+{
+    if (!m_context->bitmap)
+        return false;
+    return m_context->bitmap->set_nonvolatile(was_purged);
+}
+
+bool TGAImageDecoderPlugin::decode_tga_header()
+{
+    auto& reader = m_context->reader;
+    m_context->header = TGAHeader();
+    m_context->header.id_length = reader->read_u8();
+    m_context->header.color_map_type = reader->read_u8();
+    m_context->header.data_type_code = static_cast<TGADataType>(reader->read_u8());
+    m_context->header.color_map_origin = reader->read_i16();
+    m_context->header.color_map_length = reader->read_i16();
+    m_context->header.color_map_depth = reader->read_u8();
+    m_context->header.x_origin = reader->read_i16();
+    m_context->header.y_origin = reader->read_i16();
+    m_context->header.width = reader->read_u16();
+    m_context->header.height = reader->read_u16();
+    m_context->header.bits_per_pixel = reader->read_u8();
+    m_context->header.image_descriptor = reader->read_u8();
+
+    return true;
```

Comment:
Doesn't returning true here mean that any image decoder plugin sniffed after targa decoding will never be used? Is there any way to verify that this header contains valid data before ImageDecoder::try_create selects this plugin?

Reply:
You're right. I have added some validations to prevent this.